cpp/benchmarks/common/cudf_tpch_datagen/tpch_datagen.cpp

```diff
@@ -0,0 +1,1003 @@
+    return cudf::round(col->view(), 2);
+  }();
+
+  // Get the orderdate column from the `l_base` table
+  auto const ol_orderdate_ts = l_base->get_column(1);
+
+  // Generate the `l_shipdate` column
+  auto l_shipdate_ts = [&]() {
+    auto const l_shipdate_rand_add_days =
+      generate_random_numeric_column<int8_t>(1, 121, l_num_rows, stream, mr);
+    return add_calendrical_days(
+      ol_orderdate_ts.view(), l_shipdate_rand_add_days->view(), stream, mr);
+  }();
+
+  // Generate the `l_commitdate` column
+  auto l_commitdate_ts = [&]() {
+    auto const l_commitdate_rand_add_days =
+      generate_random_numeric_column<int8_t>(30, 90, l_num_rows, stream, mr);
+    return add_calendrical_days(
+      ol_orderdate_ts.view(), l_commitdate_rand_add_days->view(), stream, mr);
+  }();
+
+  // Generate the `l_receiptdate` column
+  auto l_receiptdate_ts = [&]() {
+    auto const l_receiptdate_rand_add_days =
+      generate_random_numeric_column<int8_t>(1, 30, l_num_rows, stream, mr);
+    return add_calendrical_days(
+      l_shipdate_ts->view(), l_receiptdate_rand_add_days->view(), stream, mr);
+  }();
+
+  // Define the current date as per clause 4.2.2.12 of the TPC-H specification
+  constexpr cudf::size_type current_date_days_since_epoch = 9'298;
+  auto current_date =
+    cudf::timestamp_scalar<cudf::timestamp_D>(current_date_days_since_epoch, true);
+  auto current_date_literal = cudf::ast::literal(current_date);
+
+  // Generate the `l_returnflag` column
+  // if `l_receiptdate` <= current_date then "R" or "A" else "N"
+  auto l_returnflag = [&]() {
+    auto const col_ref = cudf::ast::column_reference(0);
+    auto const pred =
+      cudf::ast::operation(cudf::ast::ast_operator::LESS_EQUAL, col_ref, current_date_literal);
+    auto const binary_mask =
+      cudf::compute_column(cudf::table_view({l_receiptdate_ts->view()}), pred, stream, mr);
+
+    auto const multiplier =
+      generate_repeat_sequence_column<int8_t>(2, false, l_num_rows, stream, mr);
+    auto const ternary_mask   = cudf::binary_operation(binary_mask->view(),
+                                                     multiplier->view(),
+                                                     cudf::binary_operator::MUL,
+                                                     cudf::data_type{cudf::type_id::INT8},
+                                                     stream,
+                                                     mr);
+    auto const indices        = cudf::test::fixed_width_column_wrapper<int8_t>({0, 1, 2}).release();
+    auto const keys           = cudf::test::strings_column_wrapper({"N", "A", "R"}).release();
+    auto const gather_map     = cudf::table_view({indices->view(), keys->view()});
+    auto const gathered_table = cudf::gather(
+      gather_map, ternary_mask->view(), cudf::out_of_bounds_policy::DONT_CHECK, stream, mr);
+    return std::make_unique<cudf::column>(gathered_table->get_column(1));
```

Comment:
Copy could be avoided by declaring non-const and releasing columns from table and return only required column.